### ISSUE TITLE
[FEATURE] We need better testing + communication

### ISSUE THREAD

**Issue by rijk** [framer/motion#1159]
**Is your feature request related to a problem? Please describe.**
Over the last months, I've experienced regressions in Framer Motion on an almost version by version basis. Minor versions frequently break existing implementations, while long standing issues remain unfixed. Since we decided to use Motion as a basis for our application, it has now become a major dependency to the point where if there's a blocking issue, we cannot release an update to the app, while we are powerless to improve the situation. As you can probably imagine, this has created quite some anxiety on our part. 

I don't believe this aligns with your goals either, as Framer seems to want to position itself as a tool to kickstart both design and development of professional applications; I'm guessing this has also been the original motivation to release `framer-motion` as open source. This is why I think it is important to address this issue.

**Describe the solution you'd like**
I would like to kickstart a discussion of what we can do to improve this situation. One way might be to further improve the test coverage/depth, but I'm not sure, you might have a better eye on it than I do from the outside. 

I want to help out any way that is welcome, I'm doing my best right now to document issues but am not getting much response on that, if at all. So if this is not the way, please let me know if there's a different way you'd like me/the community to contribute. I'd take the time to contribute to the tests/code, but obviously only if that is wanted and the work will not be tossed out. The lack of response to most of the issues I've posted so far (which often take a lot of effort to reproduce and document) makes me hesitant in that regard.

Lastly I want to emphasize that I love the library, I think it's the best web animation framework out there. That is why I am invested enough in it to write this, I hope this yields a constructive dialog.

**Comment by mattgperry** [comment]
Hi Rijk

I agree with the overall gist. I would also like better testing and communication. Both of these take time and I currently don't have enough of that. That's the underlying problem.

There's essentially two products here, Framer and Framer Motion. Currently, no one is paying for Framer Motion. So, when you file a bug report, and I'm responsible for another report within Framer, yours is at the back of the queue, where it stays until there's no more queue. I can assure you I'd love nothing more than to tend to my little garden and close out all the bugs but ultimately this isn't the product that is paying my bills.

What you've been experiencing over the past couple months, is we've been pushing where and how we use layout animations within Framer. As a result there's been movement within this code and where we've been fixing things in Framer we might end up breaking them in your code.

The good news is that despite appearances, testing has never been better. Because of the complexities involved in testing it, until a couple of months ago layout animations have been the one missing spot in our test coverage. Right now I just ran the test suite and it took five minutes to run all the new tests that have been added recently. So although I might have broken something in your implementation, we're increasing this testing suite and as we address more bugs and edge cases we should naturally create a more stable implementation. But we don't have a test for every single thing right now so it'll take time and more code movement.

I don't agree that you are powerless in this situation. If you are getting as much value from this as you say then you could go in and fix these problems yourself, adding a test to ensure they don't happen again. It might be that I don't have time to come in and investigate every issue right now but bug fix PRs get merged quite quickly.

**Comment by rijk** [comment]
Hi Matt, thanks for taking the time to reply in depth! Good to know PRs are welcome, I'll dive in.

Curious though, if I understand correctly Framer doesn't budget time for your maintenance of this library? Maybe you could set up a GitHub sponsor account. My app is still in its infancy but I don't mind at all paying for this library as soon as I get some revenue.

**Comment by mattgperry** [comment]
In some respects all my time is budgeted for Motion but practically speaking if there’s user-facing issues in the product this will take priority. The problem with layout animations as you’re keenly aware is that they’re either correct or drastically wrong, so when these go wrong they’re a very high priority too.

**Comment by rijk** [comment]
Hey Matt, I've been struggling to get my local clone of the package running inside my project (which I need for debugging). How do you do this? 

Things I've tried unsuccessfully:

- Running `npm link ../my-motion-clone` from my project folder -> I get the following error:
```sh
% npm link ../../../Code/React/motion -f
npm WARN using --force I sure hope you know what you are doing.
npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!
npm ERR! code EEXIST
npm ERR! path /Users/Rijk/Code/React/motion/node_modules/.bin/yarn-deduplicate
npm ERR! Refusing to delete /Users/Rijk/Code/React/motion/node_modules/.bin/yarn-deduplicate: containing path /Users/Rijk/Code/React/motion/node_modules/yarn-deduplicate isn't under npm's control
npm ERR! File exists: /Users/Rijk/Code/React/motion/node_modules/.bin/yarn-deduplicate
```

- Running `npm i ../my-motion-clone` from my project folder -> I get the following error:

![](https://aws1.discourse-cdn.com/meteor/original/3X/0/d/0d1584f96cefbf2a100d478affe6a2136a8b1fdc.png)

- Trying to fix that error by linking my project's React and React DOM to my motion clone folder; it then refuses to build:

```zsh
[2/4] 🚚  Fetching packages...
[########################################################################################################################---------------------------------------------] 952/1307events.js:187
      throw er; // Unhandled 'error' event
      ^

Error: EISDIR: illegal operation on a directory, read
Emitted 'error' event on ReadStream instance at:
    at internal/fs/streams.js:167:12
    at FSReqCallback.wrapper [as oncomplete] (fs.js:470:5) {
  errno: -21,
  code: 'EISDIR',
  syscall: 'read'
}
make: *** [node_modules/.yarn-integrity] Error 1
```

Any tips? Apologies if this is a stupid question.

**Comment by rijk** [comment]
Never mind, got it working using `npm link ../my-motion-clone` — I was using NPM v6 but after switching to NPM v7 the yarn-deduplicate error was gone.

**Comment by mattgperry** [comment]
FYI you are probably best branching from `fix/jump`

**Comment by rijk** [comment]
Good to know, thanks.

**Comment by rijk** [comment]
@mattgperry could you tell me what's going on right now with the `fix/jump` vs `feature/perfect-projection` branch (v4.1.18 vs v4.1.11 vs v5.0.0) which are developed in parallel? I'd like to contribute but it is all very hard to follow from the outside. Have you rewritten the layout projection system?

**Comment by mattgperry** [comment]
Sure, yeah I would ignore the second branch for now. Anything that is urgent you can carry on as you were like with `fix/jump`, I've probably missed some PRs but can take a look and publish a beta for you. 

The projection system is largely similar in `feature/perfect-projection` but the crucial change is it's been extracted as (essentially) a stand-alone, imperative library that can be tested in plain HTML files without React or animations. The situation as it was, was untenable, as you've probably experienced it's hard to replicate issues and genuinely in terms of my sanity I couldn't develop any further with it. So we've started afresh putting testing at the forefront.

Currently this the second week of a 6-8 week rewrite so it's early now but there's already some major improvements especially around page and element scroll, that are fully supported by tests.

**Comment by rijk** [comment]
Ah neat, that sounds like a good improvement! I'll keep an eye on the branch and sit tight for now. Would any fixes that are merged in `fix/jump` make it to the new branch as well, now that's effectively a dead end? And are new Sandboxes of layout projection issues like #1190 still useful, or not really right now?

**Comment by mattgperry** [comment]
Hopefully many of these fixes will be redundant but I will run through the branch. But the sandboxes are definitely helpful. 

One thing that was causing issues was the logic around the rebasing as different parts of the tree were using measurements based on the viewport being in different places. This has all been removed now in favour of page-relative measurements so many paths are simplified

**Comment by rijk** [comment]
> This has all been removed now in favour of page-relative measurements so many paths are simplified

Interesting. Have you read/considered the ideas in #1027? In that issue I was suggesting a way to define layout boundaries, sublayouts if you will, so that parent layout changes don't affect that part of the tree (and vice versa). An example of this would be a main/sidebar layout; as long as you make sure the `<motion.main>` element is transformed (with maybe just a `<motion.div>` straight under it for scale inversion), all their childs (which can be a lot) shouldn't need to be transformed at all. What is your take on that?
